hogben/utils.py

```diff
@@ -163,77 +157,182 @@ def fisher(qs: list[np.ndarray],
     `layer1.thick.importance = 2` in `refnx` and `refl1d` respectively.
 
-    Args:
+    Attributes:
         qs: The Q points for each model.
         xi: The varying model parameters.
         counts: incident neutron counts corresponding to each Q value.
         models: models to calculate gradients with.
         step: step size to take when calculating gradient.
-    Returns:
-        numpy.ndarray: Fisher information matrix for given models and data.
+        g: The Fisher information matrix
+        min_eigenval: The minimum eigenvalue associated with the Fisher
+         information matrix
 
     """
-    n = sum(len(q) for q in qs)  # Number of data points.
-    m = len(xi)  # Number of parameters.
-    J = np.zeros((n, m))
-
-    # There is no information if there is no data.
-    if n == 0:
-        return np.zeros((m, m))
-
-    # Calculate the gradient of model reflectivity with every model parameter
-    # for every model data point.
-    for i, parameter in enumerate(xi):
-        old = parameter.value
-
-        # Calculate reflectance for each model for first part of gradient.
-        x1 = parameter.value = old * (1 - step)
-        y1 = np.concatenate([reflectivity(q, model)
-                             for q, model in list(zip(qs, models))])
-
-        # Calculate reflectance for each model for second part of gradient.
-        x2 = parameter.value = old * (1 + step)
-        y2 = np.concatenate([reflectivity(q, model)
-                             for q, model in list(zip(qs, models))])
-
-        parameter.value = old  # Reset the parameter.
-
-        J[:, i] = (y2 - y1) / (x2 - x1)  # Calculate the gradient.
-
-    # Calculate the reflectance for each model for the given Q values.
-    r = np.concatenate([reflectivity(q, model)
-                        for q, model in list(zip(qs, models))])
-
-    # Calculate the Fisher information matrix using equations from the paper.
-    M = np.diag(np.concatenate(counts) / r, k=0)
-    g = np.dot(np.dot(J.T, M), J)
-
-    # If there are multiple parameters,
-    # scale each parameter's information by its "importance".
-    if len(xi) > 1:
-        if isinstance(xi[0], refnx.analysis.Parameter):
-            lb = np.array([param.bounds.lb for param in xi])
-            ub = np.array([param.bounds.ub for param in xi])
-
-        elif isinstance(xi[0], bumps.parameter.Parameter):
-            lb = np.array([param.bounds.limits[0] for param in xi])
-            ub = np.array([param.bounds.limits[1] for param in xi])
-
-        # Scale each parameter with their specified importance,
-        # scale with one if no importance was specified.
-        importance_array = []
-        for param in xi:
-            if hasattr(param, "importance"):
-                importance_array.append(param.importance)
-            else:
-                importance_array.append(1)
-        importance = np.diag(importance_array)
+
+    def __init__(self,
+                 qs: list[np.ndarray],
```

Comment:
Am just wondering, for multiple models, are these `list[list[np.ndarray]]]` etc?

Reply:
The xi list is still the same type for multiple models, the parameters of model 2 are just appended to Xi.